Add support for compact and masked watersheds

tlnagy opened the pull request:
This PR adds support for the compact watershed algorithm (https://www.tu-chemnitz.de/etit/proaut/publications/cws_pSLIC_ICPR.pdf). This algorithm is nice because it results in more uniform watershed basins as you increase the `compactness` parameter. 

I still need to do the following before I think it should be merged:

- [x] Add tests
- [x] More docs (and reference!)
- [x] Benchmarking (thanks to https://github.com/scikit-image/scikit-image/issues/2636 for speed enhancements)

Example segmentation following the coin example from the docs:

| Simple Watershed  | Compact watershed  |
|---|---|
| ![simple](https://user-images.githubusercontent.com/1661487/57992421-e6dde500-7a68-11e9-8ee5-3111fbf73a53.png) | ![compact](https://user-images.githubusercontent.com/1661487/57992418-e1809a80-7a68-11e9-955d-7a58e5dba775.png) |


closes #34

tlnagy committed:
add support for compact watershed

tlnagy commented:
~~Looks like my PR introduces quite a few allocations which are causing some slowdowns. Not sure where they are creeping in just yet.~~ Looks like I had some parameterization problems, the updated PR has the same performance as master

### Simple watershed

Master

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.24 MiB
  allocs estimate:  119
  --------------
  minimum time:     41.722 ms (0.00% GC)
  median time:      42.512 ms (0.00% GC)
  mean time:        43.070 ms (0.92% GC)
  maximum time:     48.598 ms (5.49% GC)
  --------------
  samples:          117
  evals/sample:     1
```

This PR

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.37 MiB
  allocs estimate:  119
  --------------
  minimum time:     44.658 ms (0.00% GC)
  median time:      45.398 ms (0.00% GC)
  mean time:        45.813 ms (0.92% GC)
  maximum time:     48.927 ms (5.84% GC)
  --------------
  samples:          110
  evals/sample:     1
```

tlnagy committed:
fix parameterization in priority queue

tlnagy committed:
add support for masked watersheds

These are useful if you don't want to partition the entire image, only
the regions labeled true in the mask

timholy commented:
Sorry for the delay here. It will take me a little while to read the paper so that I can understand this. Once you get eager to have this merged strip the WIP and ping me.

tlnagy commented:
The TL;DR on the compact watershed algorithm is that, instead of assigning the final label as a pixel is pushed into the queue, it updates the label every time it encounters that position if the new label is more optimal than the previous label. More optimal in this case is a linear combination of the euclidean distance between the seed and the pixel and the typical brightness of the pixel. It doesn't lock in the final labelling until the pixels come off the queue, this lets it find better labelings later and replace initial labelings, which cannot happen with standard watershed.

I hope that helps explain what happens.

tlnagy committed:
make sure computed compact value is the right type

timholy commented on src/watershed.jl:
```diff
@@ -18,18 +22,26 @@ Parameters:
 -    markers        = An array (same size as img) with each region's marker assigned a index starting from 1. Zero means not a marker.
                       If two markers have the same index, their regions will be merged into a single region.
                       If you have markers as a boolean array, use `label_components`.
+- compactness       = Use the compact watershed algorithm with the given compactness parameter. Larger values lead to more regularly
```
Perhaps add a citation?

timholy commented on src/watershed.jl:
````diff
@@ -1,14 +1,18 @@
 """
 ```
-segments                = watershed(img, markers)
+segments                = watershed(img, markers; compactness)
````
Add a description of `mask`?

timholy commented on src/watershed.jl:
```diff
@@ -18,18 +22,26 @@ Parameters:
-function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where {T<:Images.NumberLike, S<:Integer, N}
+function watershed(img::AbstractArray{T, N},
+                   markers::AbstractArray{S,N};
+                   mask::AbstractArray{Bool, N}=fill(true, size(img)),
```
```suggestion
                   mask::AbstractArray{Bool, N}=fill(true, axes(img)),
```

tlnagy replied:
Fixed.

timholy commented on src/watershed.jl:
```diff
@@ -1,14 +1,18 @@
+    source::CartesianIndex{N}
 end
-isless(a::PixelKey{T}, b::PixelKey{T}) where {T} = (a.val < b.val) || (a.val == b.val && a.time_step < b.time_step)
+isless(a::PixelKey{T, N}, b::PixelKey{T, N}) where {T, N} = (a.val < b.val) || (a.val == b.val && a.time_step < b.time_step)
```
I'm actually unsure why we even add the type parameter at all. Is it bad to potentially compare `PixelKey`s with different `val` types?

tlnagy replied:
I decided that comparing different values is fine. We only need dimensional parameter so we don't accidentally compare pixels with different dimensions

timholy replied:
In my comment I meant adding the type parameter to this method, not to the struct. Meaning it could be defined `isless(a::PixelKey, b::PixelKey) = ...` But if you need the `N` to match you can either reorder the parameters or use `PixelKey{S,N}` and `PixelKey{T,N}`.

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +51,61 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
+        segments_current = segments[curr_idx]
+
+        # If we're using the compact algorithm, we need assign grouping for a given location
+        # when it comes off the queue since we could have found a better suited watershed later.
```
```suggestion
        # when it comes off the queue since we could find a better suited watershed later.
```
(just making the tenses consistent, if this isn't the right change ignore it)

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +51,61 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
+                    new_value = img_current
+                else
+                    # in the compact algorithm case, we don't set the grouping at push-time and calculate
+                    # a weighted value based on the
```
this seems incomplete or confusing. Do we NOT set the grouping at push-time and also NOT calculate a weighted value...? Or is calculating the weighted value something we do INSTEAD?

timholy reviewed (dismissed):
Happy to merge this whenever you want.

tlnagy committed:
remove unnecessary type information and casting

Unnecessarily casting in the compact algorithm can lead to representation
errors with Fixed Point Numbers

tlnagy committed:
add tests for the compact watershed algorithm

tlnagy commented:
I rebased, fixed the PR based on your suggestions, fixed a representation error bug with segmenting fixed point number images, and added tests. 

I think this is ready to merge once the tests pass.

tlnagy commented:
@timholy Whenever you get a chance, could you merge this and tag a new release? Thanks!

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +52,63 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
+
+                # if this position is in the queue and we're using the compact algorithm, we need to replace
+                # its watershed if we find one that it better belongs to
+                if j in keys(pq) && compact
```
Check `compact` first since that will be quick

timholy commented on src/watershed.jl:
```diff
@@ -1,14 +1,18 @@
-struct PixelKey{CT}
-    val::CT
+struct PixelKey{N}
+    val
```
Removing the type-parameter here seems likely to cause a performance hit: the priority queue needs to order its items, and the `isless` method accesses `val`. What was the motivation for removing it rather than just adding a second parameter `N`? Was it causing problems? If so can we address this in another way?

timholy reviewed (commented):
Looks great! Any chance we could see fully up-to-date benchmarks, or has it not changed from what you posted?

I did suggest a couple of changes that seem likely to yield performance improvements.

johnnychen94 commented:
Reminder: retrigger the test after https://github.com/JuliaImages/ImageSegmentation.jl/pull/44 being merged. (Done)

tlnagy committed:
restore parameterization of pixelkey on type

but force float usage when running compact algorithm to prevent
representation errors

tlnagy commented:
I updated this PR by adding back the type paramerization in PixelKey. This fixes the performance regressions, it's basically back to the original speed. The compact algorithm now always uses floats to avoid the representation error when accumulating using fixed point numbers. @timholy and @johnnychen94, I believe this is ready to merge.

This PR:

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.44 MiB
  allocs estimate:  121
  --------------
  minimum time:     35.417 ms (0.00% GC)
  median time:      39.968 ms (0.00% GC)
  mean time:        39.711 ms (0.11% GC)
  maximum time:     42.989 ms (0.00% GC)
  --------------
  samples:          126
  evals/sample:     1
```

Master:

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.24 MiB
  allocs estimate:  119
  --------------
  minimum time:     33.161 ms (0.00% GC)
  median time:      36.747 ms (0.00% GC)
  mean time:        36.770 ms (0.11% GC)
  maximum time:     41.609 ms (0.00% GC)
  --------------
  samples:          136
  evals/sample:     1
```

tlnagy commented:
Not sure why Travis isn't showing up under checks, but the build passed: https://travis-ci.org/github/JuliaImages/ImageSegmentation.jl/builds/672668131

tlnagy commented:
Also this should fix #34

tlnagy committed:
convert to float in compact case before enqueuing

johnnychen94 commented on src/watershed.jl:
```diff
@@ -18,18 +22,27 @@ Parameters:
+    compact = compactness > 0.0
     segments = copy(markers)
-    pq = PriorityQueue{CartesianIndex{N}, PixelKey{T}}()
+    pq = PriorityQueue{CartesianIndex{N}, PixelKey{compact ? Float64 : T, N}}()
```
Instead of hardcoded `Float64`, we can use `floattype(T)`, which is almost equivalent to `float(T)`(https://github.com/JuliaMath/FixedPointNumbers.jl/issues/127). If colorant wrapper is not wanted, then it could be `floattype(eltype(T))`.
Other usages of `Float64` can be changed accordingly.

timholy replied:
I would also define this as a standalone variable
```
PK = PixelKey{floattype(eltype(T)),N}
pq = PriorityQueue{CartesianIndex{N}, PK}
```
and then use `PK(img[i], time_step, j)` below.

johnnychen94 commented on test/watershed.jl:
```diff
@@ -17,6 +17,19 @@ using ImageFiltering
+    mask = trues(size(img))
+    mask[60:70, 60:70] .= false
+
+    result = watershed(img, markers, compactness=10.0, mask=mask)
```
Does `watershed(Gray.(img), markers, compactness=10.0, mask=mask)` works normally, and how about `watershed(Gray.(ing), markers)`?

johnnychen94 reviewed (commented):
Not familiar with the algorithm details, but it generally looks good to me.

I find it doesn't have example codes in the docstring, can you add one for it?

> Not sure why Travis isn't showing up under checks

It seems to have been broken for about a month in the dark time across all repos.

johnnychen94 commented on src/watershed.jl:
```diff
@@ -18,18 +22,27 @@ Parameters:
+function watershed(img::AbstractArray{T, N},
+                   markers::AbstractArray{S,N};
+                   mask::AbstractArray{Bool, N}=fill(true, axes(img)),
+                   compactness::Float64 = 0.0) where {T<:Images.NumberLike, S<:Integer, N}
```
```suggestion
                   compactness::Real = 0.0) where {T<:Images.NumberLike, S<:Integer, N}
```

`Float64` could be problematic:

```julia
julia> function foo(; x::Float64)
       @show x
       end
foo (generic function with 1 method)

julia> foo(x=1.0)
x = 1.0
1.0

julia> foo(x=1.0f0)
ERROR: TypeError: in keyword argument x, expected Float64, got Float32
Stacktrace:
 [1] top-level scope at REPL[21]:1
```

tlnagy replied:
I loosened the type signature.

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +52,63 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
                 if segments[j] == 0
                     segments[j] = markers[i]
-                    enqueue!(pq, j, PixelKey(img[i], time_step))
+                    enqueue!(pq, j, PixelKey(compact ? Float64.(img[1]) : img[i], time_step, j))
```
Is the `img[1]` a bug? Should it be `img[i]`? See also the `PK` suggestion above.

tlnagy replied:
Good catch...

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +52,63 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
+                # its watershed if we find one that it better belongs to
+                if compact && j in keys(pq)
+                    elem = pq[j]
+                    new_elem = PixelKey(new_value, time_step, curr_elem.source)
```
Should also make this `PK(new_value, time_step, curr_elem.source)`.

timholy commented on src/watershed.jl:
```diff
@@ -39,22 +52,63 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
+                    end
+                else
+
+                    pq[j] = PixelKey(new_value, time_step, curr_elem.source)
```
Likewise

timholy reviewed (commented):
Sorry for the delay. I like @johnnychen94's comments, if we can address the last bit let's get this merged!

tlnagy committed:
remove strict float64 casting and loosen type of compactness

tlnagy committed:
test watershed options separately

tlnagy committed:
add simple example to watershed docstring

tlnagy commented:
@johnnychen94 and @timholy I finally got around to making all the suggested changes. Hope this is finally ready for a merge!

johnnychen94 reviewed (approved):
LGTM, will merge in three days if there's no further comments

tlnagy commented:
Bump @johnnychen94 and @timholy. Would be great to have this merged!

johnnychen94 commented:
Thanks for the work and patience @tlnagy !

tlnagy commented:
Could you tag a new version when you get a chance? Thanks!

timholy commented:
Thanks @tlnagy, and @johnnychen94 for merging!
